### PR TITLE
FXC-3603: security fixes in github actions as suggested by zizmor

### PR DESCRIPTION
**Pull request by mahlau-flex**
Fixed all (except one) of the many security vulnerabilities found by zizmor (12 informational, 0 low, 19 medium, 19 high). While I am pretty sure that the functionality remains the same, we will need to see (an carefully review) if something broke during these changes.

The one issue I could not change is the following (status informational):
```
info[use-trusted-publishing]: prefer trusted publishing for authentication
   --> .github/workflows/tidy3d-python-client-release.yml:121:9
    |
119 |       run: |
    |       --- this step
120 |         python -m build
121 |         python -m twine upload --repository pypi dist/*
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this command
    |
    = note: audit confidence → High
```
I have suppressed the zizmor warning for now. In the long run it would probably be best to use the newer authentication of pypi releases (according to Claude, one can just register a github workflow in pypi instead of using the legacy token system).

Additionally, I have added a pre-commit check for zizmor as well. This might seem a bit excessive, but runs extremely quickly so I think it will not bother anyone.

Lastly, I changed the configuration of zizmor itself in the github actions. I noticed that in the last PR, the job completed successfully, even though it found a security issue. This is, because it uploads the results to github advanced security, which then makes a comment showing the issue. But, this does not prevent merging and is not clearly visibly. With the new system, the checks themselves should (hopefully) fail.

<h2>Greptile Overview</h2>

Updated On: 2025-10-09 19:03:13 UTC

<h3>Summary</h3>
This PR implements comprehensive security hardening for GitHub Actions workflows based on findings from the zizmor security scanner, which identified 50 security vulnerabilities (12 informational, 19 medium, 19 high) across the CI/CD pipeline. The changes follow GitHub Actions security best practices without altering functionality.

Key security improvements include:

1. **Permission Management**: Moving from broad permissions to the principle of least privilege by setting global permissions to `contents: read` and granting specific permissions only where needed (e.g., `contents: write` for jobs that push changes)

2. **Action Security**: Pinning action versions to specific commit SHAs instead of floating tags to prevent supply chain attacks where malicious code could be injected through compromised action updates

3. **Credential Protection**: Adding `persist-credentials: false` to checkout actions to prevent GitHub tokens from being accessible to subsequent workflow steps that don't need them

4. **Script Injection Prevention**: Moving GitHub context expressions (like PR titles and branch names) to environment variables to prevent potential script injection attacks

5. **Proactive Security**: Adding zizmor as both a dependency in `pyproject.toml` and a pre-commit hook to catch future security issues during development

The changes span 8 workflow files covering testing, releases, documentation sync, and daily operations. One informational issue regarding PyPI trusted publishing was acknowledged but left unaddressed, with plans to migrate to the newer authentication system in the future. The implementation maintains all existing functionality while significantly reducing the attack surface of the CI/CD pipeline.
<h3>Important Files Changed</h3>

<details><summary>Changed Files</summary>

| Filename | Score | Overview |
|----------|-------|----------|
| `.github/workflows/tidy3d-python-client-tests.yml` | 4/5 | Comprehensive security hardening with action pinning, credential protection, and zizmor integration |
| `.github/workflows/tidy3d-python-client-release.yml` | 4/5 | Security improvements for release workflow with permission scoping and action pinning |
| `.github/workflows/tidy3d-python-client-update-lockfile.yml` | 4/5 | Security hardening for lockfile update workflow with SHA pinning and permission controls |
| `.github/workflows/tidy3d-python-client-daily.yml` | 4/5 | Permission restrictions and explicit secret passing for daily workflow security |
| `.github/workflows/tidy3d-docs-sync-readthedocs-repo.yml` | 5/5 | Clean security improvements with minimal permission grants for documentation sync |
| `.github/workflows/tidy3d-python-client-submodules-test.yml` | 5/5 | Simple security hardening with read-only permissions and credential protection |
| `.github/workflows/tidy3d-python-client-develop-cli.yml` | 5/5 | Standard security improvements with action pinning and permission restrictions |
| `.pre-commit-config.yaml` | 4/5 | Addition of zizmor pre-commit hook for proactive security scanning |
| `pyproject.toml` | 5/5 | Clean addition of zizmor dependency for security tooling integration |

</details>
<h3>Confidence score: 4/5</h3>

- This PR is safe to merge with careful monitoring of CI/CD pipeline functionality
- Score reflects the comprehensive nature of security changes across critical workflow files that require validation
- Pay close attention to all workflow files to ensure CI/CD operations continue functioning correctly

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant Dev as "Developer"
    participant GitHub as "GitHub Actions"
    participant PyPI
    participant ReadTheDocs as "ReadTheDocs Mirror"
    participant Zizmor as "Zizmor Security Tool"
    
    User->>Dev: "Request security fixes for GitHub Actions"
    Dev->>Zizmor: "Run security audit on .github/workflows/*"
    Zizmor-->>Dev: "Report 12 informational, 19 medium, 19 high vulnerabilities"
    
    Dev->>GitHub: "Update workflow files with security fixes"
    Note over Dev,GitHub: Pin action versions to specific commit hashes<br/>Add persist-credentials: false<br/>Set minimal permissions<br/>Update to latest action versions
    
    Dev->>GitHub: "Add zizmor pre-commit hook"
    GitHub->>Zizmor: "Run zizmor check on workflow files"
    Zizmor-->>GitHub: "Security validation results"
    
    Dev->>GitHub: "Configure zizmor to fail CI on security issues"
    Note over GitHub: Previously uploaded to Advanced Security<br/>Now fails the check directly
    
    GitHub->>PyPI: "Release workflow with suppressed trusted publishing warning"
    Note over GitHub,PyPI: Legacy token authentication maintained<br/>zizmor warning suppressed
    
    GitHub->>ReadTheDocs: "Sync documentation with security fixes"
    
    Dev->>User: "Security fixes complete (except trusted publishing)"
    Note over User,Dev: All issues fixed except one informational<br/>about using trusted publishing for PyPI
```

### CODE REVIEW

**Comment by mahlau-flex** [comment]
I have renamed the title of this PR to link to the Zizmor jira task. Are we supposed to link multiple PRs to a single jira issue or should we create a new one everytime?

File: .github/workflows/tidy3d-python-client-tests.yml [hunk 1]
```diff
@@ -521,7 +553,7 @@ jobs:
 
     - name: install-project
       shell: bash
-      if: ${{ matrix.platform }} != "windows-latest"
+      if: matrix.platform != 'windows-latest'
```
Comment (mahlau-flex) [inline .github/workflows/tidy3d-python-client-tests.yml]: One thing I should add: This if-case was always true in the old version due to how github actions works. The new version implements the correct check. It is worthwile to discuss, if we want to remove the if-clause completely, because previously we seemed to test on windows anyways.

Comment (yaugenst-flex) [reply .github/workflows/tidy3d-python-client-tests.yml]: @daquinteroflex do you know what the reasoning there was? I'd tend toward just removing it, as @mahlau-flex is suggesting.

File: .github/workflows/tidy3d-python-client-tests.yml [hunk 2]
```diff
@@ -23,7 +23,6 @@ on:
 
 permissions:
   contents: read
-  pull-requests: write
```
Comment (yaugenst-flex) [inline .github/workflows/tidy3d-python-client-tests.yml:26]: Could you explain this change? Wouldn't this mean that the scope is now missing from `GITHUB_TOKEN`, but there are workflow steps like `check-current-approval-status` that do require access to the PR? Would `pull-requests: read` do the trick?

**Review by yaugenst-flex — COMMENTED** [review]
Thanks @mahlau-flex this is great! Besides what was already commented LGTM. I'm not too familiar with the details but as long as everything keeps running it looks fine to me.

File: .github/workflows/tidy3d-python-client-tests.yml (hunk 2 above)
Comment (mahlau-flex) [reply .github/workflows/tidy3d-python-client-tests.yml:26]: The problem with this line was, that not all of the jobs need the write permission. Zizmor declared this as a "too broad" permission, so I moved the write permission only to the jobs that actually need it (see line 421 and 520).

**Review by yaugenst-flex — APPROVED** [review]

**Comment by mahlau-flex** [comment]
hm, it seems that some tests are actually failing here. It looks like these are the discussed windows test cases. I will investigate how to fix this

File: .github/workflows/tidy3d-python-client-develop-cli.yml [hunk 3]
```diff
@@ -58,7 +62,7 @@ jobs:
       #  -----  install & configure poetry  -----
       #----------------------------------------------
     - name: Install Poetry
-      uses: snok/install-poetry@v1
+      uses: snok/install-poetry@76e04a911780d5b312d89783f7b1cd627778900a  # v1.4.1
       with:
         version: 1.8.2
```
Comment (daquinteroflex) [inline .github/workflows/tidy3d-python-client-develop-cli.yml]: Will this work with poetry 2 as we've upgraded the python client tests to that?

**Comment by daquinteroflex** [comment]
Hi @mahlau-flex thanks! Could you share some github actions run links where you test each workflow bar the release one?

File: .github/workflows/tidy3d-python-client-tests.yml [hunk 4]
```diff
@@ -581,29 +613,39 @@ jobs:
...
     steps:
       - name: check-passing-remote-tests
+        env:
+          NEEDS_LOCAL_TESTS_RESULT: ${{ needs.local-tests.result }}
+          NEEDS_REMOTE_TESTS_RESULT: ${{ needs.remote-tests.result }}
+          NEEDS_LINT_RESULT: ${{ needs.lint.result }}
+          NEEDS_VERIFY_SCHEMA_CHANGE_RESULT: ${{ needs.verify-schema-change.result }}
+          NEEDS_LINT_COMMIT_MESSAGES_RESULT: ${{ needs.lint-commit-messages.result }}
+          NEEDS_LINT_BRANCH_NAME_RESULT: ${{ needs.lint-branch-name.result }}
+          NEEDS_ZIZMOR_RESULT: ${{ needs.zizmor.result }}
```
Comment (daquinteroflex) [inline .github/workflows/tidy3d-python-client-tests.yml]: Nice to consolidate!

File: .github/workflows/tidy3d-python-client-tests.yml (hunk 1 above)
Comment (daquinteroflex) [reply .github/workflows/tidy3d-python-client-tests.yml]: Hmm interesting, how did you find out it was always true? If so feel free to remove

File: .github/workflows/tidy3d-python-client-tests.yml [hunk 5]
```diff
@@ -436,11 +465,13 @@ jobs:
...
         github.event_name == 'pull_request' &&
         !contains(github.event.pull_request.labels.*.name, 'ignore_diff_coverage')
+      env:
+        GITHUB_EVENT_PULL_REQUEST_BASE_REF: ${{ github.event.pull_request.base.ref }}
       run: |
         source ${GITHUB_WORKSPACE}/.venv/bin/activate
         git config --global --add safe.directory ${GITHUB_WORKSPACE}
         diff-cover ${GITHUB_WORKSPACE}/coverage.xml \
-          --compare-branch origin/${{ github.event.pull_request.base.ref }} \
+          --compare-branch origin/${GITHUB_EVENT_PULL_REQUEST_BASE_REF} \
```
Comment (daquinteroflex) [inline .github/workflows/tidy3d-python-client-tests.yml:487]: Thanks a lot great to finally clean this up more!

**Review by daquinteroflex — COMMENTED** [review]
Thanks a lot @mahlau-flex this is a good cleanup. Let's just do some final checking on the workflow runs and then we can merge

File: .github/workflows/tidy3d-python-client-develop-cli.yml (hunk 3 above)
Comment (mahlau-flex) [reply .github/workflows/tidy3d-python-client-develop-cli.yml]: The hash only pins the version of the github actions script. If we want to change the poetry version we can change the ```with version...``` part. Do you want to me change this to the latest poetry version?

File: .github/workflows/tidy3d-python-client-tests.yml (hunk 1 above)
Comment (mahlau-flex) [reply .github/workflows/tidy3d-python-client-tests.yml]: zizmor put a hint that this expression always evaluates to true, though this seems a complicated edge case and github actions does not seem to document this well. If I understand correctly, the "${{}}" in an if-clause are evaluated and mark the end of evaluation when done. So the result would be "windows-latest" != "windows-latest", but since evaluation is done the whole term would be treated as a non-empty string, which results in true always.

Comment (mahlau-flex) [reply .github/workflows/tidy3d-python-client-tests.yml]: Looking at the github workflow [here](https://github.com/flexcompute/tidy3d/actions/runs/18491875283/job/52687208069?pr=2882) it seems that with the corrected expression the workflow actually will always fail because then subsequent steps are missing the project. I will see to remove the expression completely such that this works again.

**Comment by mahlau-flex** [comment]
> Hi @mahlau-flex thanks! Could you share some github actions run links where you test each workflow bar the release one?

Sorry, I am not sure what you mean by that. Is it possible to manually trigger the workflows as a test? Would this break anything?

File: .github/workflows/tidy3d-python-client-tests.yml (hunk 1 above)
Comment (mahlau-flex) [reply .github/workflows/tidy3d-python-client-tests.yml]: I have removed it

**Comment by daquinteroflex** [comment]
So yeah if you manually run all bar the release action from this branch it should be fine, it's just to check all works nicely now

File: .github/workflows/tidy3d-python-client-develop-cli.yml (hunk 3 above)
Comment (daquinteroflex) [reply .github/workflows/tidy3d-python-client-develop-cli.yml]: Yeah since you're cleaning it up can you make sure they're all using the same action version and poetry version too? Ideally the latest version. Sometimes they change the action version for the poetry version I remember

**Comment by mahlau-flex** [comment]
Everything should be working now. Here are links to the successfull manual runs:
- https://github.com/flexcompute/tidy3d/actions/runs/18556073617
- https://github.com/flexcompute/tidy3d/actions/runs/18555999294
- https://github.com/flexcompute/tidy3d/actions/runs/18556024394
- https://github.com/flexcompute/tidy3d/actions/runs/18563140437

The daily run fails, but I believe this has nothing to do with the changes here, but rather that the test is doing what its supposed to be doing and alerting that we need to update a submodule: 
https://github.com/flexcompute/tidy3d/actions/runs/18563335630

**Review by daquinteroflex — APPROVED** [review]
Thanks a lot!
